blocks/autocompleters/index.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import { createBlock, getBlockTypes } from '../api';
+import BlockIcon from '../block-icon';
+
+/**
+ * @typedef {Object} CompleterOption
+ * @property {Array.<Component>} label list of react components to render.
+ * @property {Array.<String>} keywords list of key words to search.
+ * @property {*} value the value that will be passed to onSelect.
+ */
+
+/**
+ * @callback FnGetOptions
+ * @returns {Promise.<Array.<CompleterOption>>} A promise that resolves to the list of completer options.
+ */
+
+/**
+ * @callback FnAllowNode
+ * @param {Node} textNode check if the completer can handle this text node.
+ * @returns {boolean} true if the completer can handle this text node.
+ */
+
+/**
+ * @callback FnAllowContext
+ * @param {Range} before the range before the auto complete trigger and query.
+ * @param {Range} range the range of the autocomplete trigger and query.
+ * @param {Range} after the range after the autocomplete trigger and query.
+ * @returns {boolean} true if the completer can handle these ranges.
+ */
+
+/**
+ * @callback FnOnSelect
+ * @param {*} value the value of the completer option.
+ * @param {Range} range the nodes included in the autocomplete trigger and query.
+ * @param {String} query the text value of the autocomplete query.
+ */
+
+/**
+ * @typedef {Object} Completer
+ * @property {?String} className A class to apply to the popup menu.
+ * @property {String} triggerPrefix the prefix that will display the menu.
+ * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
+ * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
+ * @property {FnOnSelect} onSelect
+ */
+
+/**
+ * Returns an "completer" definition for selecting from available blocks to replace the current one.
+ * The definition can be understood by the Autocomplete component.
+ *
+ * @param  {Function} onReplace  Callback to replace the current block.
+ * @returns {Completer}          Completer object used by the Autocomplete component.
+ */
+export function blockAutocompleter( { onReplace } ) {
+	const options = getBlockTypes().map( ( blockType ) => {
+		const { name, title, icon, keywords = [] } = blockType;
+		return {
+			value: name,
+			label: [
+				<BlockIcon key="icon" icon={ icon } />,
+				title,
+			],
+			keywords: [ ...keywords, title ],
+		};
+	} );
+
+	const getOptions = () => Promise.resolve( options );
+
+	const allowContext = ( before, range, after ) => {
+		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
+	};
+
+	const onSelect = ( blockName ) => {
+		onReplace( createBlock( blockName ) );
+	};
+
+	return {
+		className: 'blocks-block-autocomplete',
+		triggerPrefix: '/',
+		getOptions,
+		allowContext,
```

Comment:
Can't understand the difference between `triggerPrefix` and  `allowContext`, could these be merged into one function?

Reply:
TLDR: not without making it an annoying API to use.

triggerPrefix is the string that preceeds the query. For example "@" in the case of the user mentions, "/" in the case of block selection and "#" in the case of hash-tags (which aren't implemented yet). In theory you could just search back to the last space or boundary and pass the whole range to the completer definition to check but that that means that ALL completers would have to convert the range into text and check the leading substring. This puts more work on the implementer of the completer which is exactly where we don't want it.

allowContext was added to solve the problem that sometimes the completer has to know what preceeds it and follows it. The majority of the time this doesn't matter but the original implementation of the autocomplete block would require that there was nothing but whitespace before and afterwards (which makes sense because it replaces the whole block). This field is completely optional and most completers won't use it.